File: typescript/src/transaction.ts
/**
 * The transaction message.
 *
 * A fixed 104-byte header followed by sections in a fixed order, each 8-byte aligned:
 *
 * ```text
 *   0  u64  slot                 64  u32  fec_set_index
 *   8  u64  parent_slot          68  u32  entry_index
 *  16  u64  rx_ts_ns             72  u32  tx_index
 *  24  u64  emit_ts_ns           76  u32  instruction_bytes_len
 *  32  [32] leader               80  u32  lookup_bytes_len
 *                                84  u32  raw_tx_len
 *                                88  u16  source_id
 *                                90  u16  flags
 *                                92  u16  account_count
 *                                94  u16  instruction_count
 *                                96  u8   signature_count
 *                                97  u8   message_version   (0xff = legacy)
 *                                98  u8   lookup_count
 *                                99  u8   required_signatures
 *                               100  u8   readonly_signed
 *                               101  u8   readonly_unsigned
 *                               102  u16  reserved
 * 104     signatures         signature_count * 64
 *         account_keys       account_count * 32
 *         recent_blockhash   32
 *         instruction_table  instruction_count * 16
 *         instruction_bytes  padded to 8
 *         lookup_table       lookup_count * 40
 *         lookup_bytes       padded to 8
 *         raw_tx             raw_tx_len, only when the flag is set
 * ```
 *
 * The fixed offsets are what make reading cheap: every field is a load at a known position and
 * every byte string is a `subarray`, never a copy. Nothing is decoded until it is asked for, so a
 * subscriber that only looks at, say, the first account key never pays for the rest.
 */

import { ProtocolError, Verification } from './protocol.js';

/** Bytes in the fixed header. */
export const TX_HEADER_LEN = 104;
/** Bytes in one instruction table entry. */
export const INSTRUCTION_ENTRY_LEN = 16;
/** Bytes in one address lookup table entry. */
export const LOOKUP_ENTRY_LEN = 40;

/** Bits in the transaction header's flag field. */
export const TxFlag = {
  /** The transaction is a simple vote. */
  IS_VOTE: 1 << 0,
  /** The FEC set's leader signature verified. */
  VERIFIED: 1 << 1,
  /** The raw transaction bytes are appended. */
  HAS_RAW_TX: 1 << 2,
  /** The FEC set had to be reconstructed from parity. */
  RECOVERED: 1 << 3,
  /** The leader was known and named in the header. */
  HAS_LEADER: 1 << 4,
} as const;

const REASON_SHIFT = 5;
const REASON_MASK = 0b11;

/** Marks a legacy (pre-versioned) message. */
export const MESSAGE_VERSION_LEGACY = 0xff;

/** One instruction, pointing into the message's shared byte region. */
export interface Instruction {
  /** Index into `accountKeys` of the program being invoked. */
  programIdIndex: number;
  /** Indices into `accountKeys`, in the order the program expects them. */
  accounts: Buffer;
  /** Opaque instruction data. */
  data: Buffer;
}

/** One address lookup table reference, carried unresolved. */
export interface AddressLookup {
  /** The lookup table account. */
  accountKey: Buffer;
  /** Indices loaded as writable. */
  writableIndexes: Buffer;
  /** Indices loaded as readonly. */
  readonlyIndexes: Buffer;
}

/** Rounds up to a multiple of 8. */
function align8(len: number): number {
  return (len + 7) & ~7;
}

/**
 * A decoded transaction.
 *
 * Backed directly by the received buffer. Accessors return `subarray` views, which means they are
 * valid only as long as the underlying buffer is — copy anything kept past the callback with
 * `Buffer.from(...)`.
 */
export class Transaction {
  private readonly signaturesAt: number;
  private readonly accountKeysAt: number;
  private readonly blockhashAt: number;
  private readonly instructionTableAt: number;
  private readonly instructionBytesAt: number;
  private readonly lookupTableAt: number;
  private readonly lookupBytesAt: number;
  private readonly rawTxAt: number;

  private constructor(readonly bytes: Buffer) {
    const need = (end: number): never => {
      throw new ProtocolError(`transaction needs ${end} bytes, have ${bytes.length}`);
    };
    let at = TX_HEADER_LEN;
    const take = (len: number): number => {
      const start = at;
      at += len;
      if (at > bytes.length) need(at);
      return start;
    };

    this.signaturesAt = take(this.signatureCount * 64);
    this.accountKeysAt = take(this.accountCount * 32);
    this.blockhashAt = take(32);
    this.instructionTableAt = take(this.instructionCount * INSTRUCTION_ENTRY_LEN);
    const instructionBytesLen = this.bytes.readUInt32LE(76);
    this.instructionBytesAt = take(align8(instructionBytesLen));
    this.lookupTableAt = take(this.lookupCount * LOOKUP_ENTRY_LEN);
    const lookupBytesLen = this.bytes.readUInt32LE(80);
    this.lookupBytesAt = take(align8(lookupBytesLen));
    this.rawTxAt = take(this.bytes.readUInt32LE(84));

    // Table entries are offsets into their own byte region. Checking them once here is what lets
    // every accessor below slice without a bounds test.
    for (let i = 0; i < this.instructionCount; i += 1) {
      const entry = this.instructionTableAt + i * INSTRUCTION_ENTRY_LEN;
      const accountsEnd = bytes.readUInt16LE(entry + 2) + bytes.readUInt16LE(entry + 4);
      const dataEnd = bytes.readUInt32LE(entry + 8) + bytes.readUInt32LE(entry + 12);
      if (accountsEnd > instructionBytesLen || dataEnd > instructionBytesLen) {
        throw new ProtocolError(`instruction ${i} points outside its byte region`);
      }
    }
    for (let i = 0; i < this.lookupCount; i += 1) {
      const entry = this.lookupTableAt + i * LOOKUP_ENTRY_LEN;
      const writableEnd = bytes.readUInt16LE(entry + 32) + bytes.readUInt16LE(entry + 34);
      const readonlyEnd = bytes.readUInt16LE(entry + 36) + bytes.readUInt16LE(entry + 38);
      if (writableEnd > lookupBytesLen || readonlyEnd > lookupBytesLen) {
        throw new ProtocolError(`lookup ${i} points outside its byte region`);
      }
    }
  }

  /** Reads a transaction, validating that every section is present. */
  static read(bytes: Buffer): Transaction {
    if (bytes.length < TX_HEADER_LEN) {
      throw new ProtocolError(`transaction header needs ${TX_HEADER_LEN} bytes, have ${bytes.length}`);
    }
    return new Transaction(bytes);
  }

  /** Slot the transaction landed in. */
  get slot(): bigint {
    return this.bytes.readBigUInt64LE(0);
  }

  /** The slot this one builds on. */
  get parentSlot(): bigint {
    return this.bytes.readBigUInt64LE(8);
  }

  /** When the first shred of the FEC set arrived, on the server's monotonic clock. */
  get rxTsNs(): bigint {
    return this.bytes.readBigUInt64LE(16);
  }

  /** When the server finished encoding this message, on the same clock as `rxTsNs`. */
  get emitTsNs(): bigint {
    return this.bytes.readBigUInt64LE(24);
  }

  /**
   * How long the server took from first shred to encoded message.
   *
   * Both stamps come from one monotonic clock, so this is meaningful; neither can be compared
   * against a local wall clock.
   */
  get pipelineNs(): bigint {
    return this.emitTsNs - this.rxTsNs;
  }

  /** Leader assigned to the slot, when the schedule knew one. */
  get leader(): Buffer | null {
    return this.has(TxFlag.HAS_LEADER) ? this.bytes.subarray(32, 64) : null;
  }

  /** Erasure set this transaction was carried in. */
  get fecSetIndex(): number {
    return this.bytes.readUInt32LE(64);
  }

  /** Index of the entry within the slot. */
  get entryIndex(): number {
    return this.bytes.readUInt32LE(68);
  }

  /** Index of the transaction within its entry. */
  get txIndex(): number {
    return this.bytes.readUInt32LE(72);
  }

  /** Which ingress source delivered the shreds. */
  get sourceId(): number {
    return this.bytes.readUInt16LE(88);
  }

  /** Raw header flags. Prefer the named accessors. */
  get flags(): number {
    return this.bytes.readUInt16LE(90);
  }

  private has(bits: number): boolean {
    return (this.flags & bits) === bits;
  }

  /** Whether this is a simple vote transaction. */
  get isVote(): boolean {
    return this.has(TxFlag.IS_VOTE);
  }

  /** Whether the FEC set had to be reconstructed from parity shreds. */
  get recovered(): boolean {
    return this.has(TxFlag.RECOVERED);
  }

  /** How much the server could vouch for this transaction. */
  get verification(): Verification {
    if (this.has(TxFlag.VERIFIED)) return Verification.Verified;
    switch ((this.flags >> REASON_SHIFT) & REASON_MASK) {
      case 1:
        return Verification.UnknownLeader;
      case 2:
        return Verification.StaleSchedule;
      default:
        return Verification.Disabled;
    }
  }

  /** Number of account keys carried in the message. */
  get accountCount(): number {
    return this.bytes.readUInt16LE(92);
  }

  /** Number of instructions. */
  get instructionCount(): number {
    return this.bytes.readUInt16LE(94);
  }

  /** Number of signatures. */
  get signatureCount(): number {
    return this.bytes.readUInt8(96);
  }

  /** Message version, or {@link MESSAGE_VERSION_LEGACY} for a legacy message. */
  get messageVersion(): number {
    return this.bytes.readUInt8(97);
  }

  /** Number of address lookup table references. */
  get lookupCount(): number {
    return this.bytes.readUInt8(98);
  }

  /** Signatures the message requires. */
  get requiredSignatures(): number {
    return this.bytes.readUInt8(99);
  }

  /** Signed accounts that are readonly. */
  get readonlySigned(): number {
    return this.bytes.readUInt8(100);
  }

  /** Unsigned accounts that are readonly. */
  get readonlyUnsigned(): number {
    return this.bytes.readUInt8(101);
  }

  /**
   * The transaction's first signature — its id. `null` if it carries none.
   *
   * Every real transaction is signed, so this is `null` only for something a leader should never
   * have put in a block. It is nullable rather than throwing because the node is a relay, not a
   * validator: it forwards what the leader signed into the shred, and refusing the frame or
   * throwing from an accessor would let a leader knock subscribers off the stream. `null` is
   * something a consumer can skip; an exception from a property read, thrown well outside whatever
   * `try` guarded the decode, is not.
   */
  get signature(): Buffer | null {
    if (this.signatureCount === 0) return null;
    return this.bytes.subarray(this.signaturesAt, this.signaturesAt + 64);
  }

  /** Every signature, in order. */
  signatures(): Buffer[] {
    const out: Buffer[] = [];
    for (let i = 0; i < this.signatureCount; i += 1) {
      const at = this.signaturesAt + i * 64;
      out.push(this.bytes.subarray(at, at + 64));
    }
    return out;
  }

  /**
   * One account key, by index. `null` if the index is past the end.
   *
   * Nullable because the index usually is not the caller's: `instruction.programIdIndex` comes off
   * the wire, and a transaction naming an index past its own account list is something a leader can
   * produce. Throwing here would turn that into an exception in the middle of a `for` loop over
   * instructions — the pattern this package's own documentation shows.
   */
  accountKey(index: number): Buffer | null {
    if (index < 0 || index >= this.accountCount) return null;
    const at = this.accountKeysAt + index * 32;
    return this.bytes.subarray(at, at + 32);
  }

  /** Every account key, in the order the message declared them. */
  accountKeys(): Buffer[] {
    const out: Buffer[] = [];
    for (let i = 0; i < this.accountCount; i += 1) {
      const at = this.accountKeysAt + i * 32;
      out.push(this.bytes.subarray(at, at + 32));
    }
    return out;
  }

  /** The blockhash the transaction was signed against. */
  get recentBlockhash(): Buffer {
    return this.bytes.subarray(this.blockhashAt, this.blockhashAt + 32);
  }

  /** One instruction, by index. `null` if the index is past the end. */
  instruction(index: number): Instruction | null {
    if (index < 0 || index >= this.instructionCount) return null;
    const entry = this.instructionTableAt + index * INSTRUCTION_ENTRY_LEN;
    const base = this.instructionBytesAt;
    const accountsOff = base + this.bytes.readUInt16LE(entry + 2);
    const dataOff = base + this.bytes.readUInt32LE(entry + 8);
    return {
      programIdIndex: this.bytes.readUInt8(entry),
      accounts: this.bytes.subarray(accountsOff, accountsOff + this.bytes.readUInt16LE(entry + 4)),
      data: this.bytes.subarray(dataOff, dataOff + this.bytes.readUInt32LE(entry + 12)),
    };
  }

  /** Every instruction, in execution order. */
  instructions(): Instruction[] {
    const out: Instruction[] = [];
    for (let i = 0; i < this.instructionCount; i += 1) {
      // In range by construction, so the `null` case cannot arise here.
      const instruction = this.instruction(i);
      if (instruction !== null) out.push(instruction);
    }
    return out;
  }

  /**
   * Every address lookup table reference, unresolved.
   *
   * Resolving these needs the lookup tables' account state, which a shred pipeline does not have.
   * A consumer that needs the resolved keys must fetch the tables itself.
   */
  lookups(): AddressLookup[] {
    const out: AddressLookup[] = [];
    for (let i = 0; i < this.lookupCount; i += 1) {
      const entry = this.lookupTableAt + i * LOOKUP_ENTRY_LEN;
      const base = this.lookupBytesAt;
      const writableOff = base + this.bytes.readUInt16LE(entry + 32);
      const readonlyOff = base + this.bytes.readUInt16LE(entry + 36);
      out.push({
        accountKey: this.bytes.subarray(entry, entry + 32),
        writableIndexes: this.bytes.subarray(
          writableOff,
          writableOff + this.bytes.readUInt16LE(entry + 34),
        ),
        readonlyIndexes: this.bytes.subarray(
          readonlyOff,
          readonlyOff + this.bytes.readUInt16LE(entry + 38),
        ),
      });
    }
    return out;
  }

  /**
   * The original transaction bytes, when the server was asked to include them.
   *
   * Off by default: the decoded form above carries the same information, and repeating the bytes
   * roughly doubles the stream's bandwidth.
   */
  get rawTx(): Buffer | null {
    if (!this.has(TxFlag.HAS_RAW_TX)) return null;
    return this.bytes.subarray(this.rawTxAt, this.rawTxAt + this.bytes.readUInt32LE(84));
  }
}

File: rust/src/transaction.rs
//! The transaction message: a fully decoded transaction laid out for zero-copy reading.
//!
//! # Layout
//!
//! A fixed 104-byte header followed by sections in a fixed order, each starting 8-byte aligned:
//!
//! ```text
//!   0  u64  slot                 64  u32  fec_set_index
//!   8  u64  parent_slot          68  u32  entry_index
//!  16  u64  rx_ts_ns             72  u32  tx_index
//!  24  u64  emit_ts_ns           76  u32  instruction_bytes_len
//!  32  [32] leader               80  u32  lookup_bytes_len
//!                                84  u32  raw_tx_len
//!                                88  u16  source_id
//!                                90  u16  flags
//!                                92  u16  account_count
//!                                94  u16  instruction_count
//!                                96  u8   signature_count
//!                                97  u8   message_version   (0xff = legacy)
//!                                98  u8   lookup_count
//!                                99  u8   required_signatures
//!                               100  u8   readonly_signed
//!                               101  u8   readonly_unsigned
//!                               102  u16  reserved
//! 104     signatures         signature_count * 64
//!         account_keys       account_count * 32
//!         recent_blockhash   32
//!         instruction_table  instruction_count * 16
//!         instruction_bytes  padded to 8
//!         lookup_table       lookup_count * 40
//!         lookup_bytes       padded to 8
//!         raw_tx             raw_tx_len, only when the flag is set
//! ```
//!
//! Fixed offsets and alignment are what make reading free: account keys and signatures are read by
//! pointing at a slice, never by walking length prefixes, and nothing is decoded until it is asked
//! for.

use crate::{
    error::{Error, Result},
    protocol::Verification,
};

/// Bytes in the fixed header.
pub const TX_HEADER_LEN: usize = 104;
/// Bytes in one instruction table entry.
pub const INSTRUCTION_ENTRY_LEN: usize = 16;
/// Bytes in one address lookup table entry.
pub const LOOKUP_ENTRY_LEN: usize = 40;
/// Marks a legacy (pre-versioned) message.
pub const MESSAGE_VERSION_LEGACY: u8 = 0xff;

const IS_VOTE: u16 = 1 << 0;
const VERIFIED: u16 = 1 << 1;
const HAS_RAW_TX: u16 = 1 << 2;
const RECOVERED: u16 = 1 << 3;
const HAS_LEADER: u16 = 1 << 4;
const REASON_SHIFT: u16 = 5;
const REASON_MASK: u16 = 0b11;

/// Rounds up to a multiple of 8.
const fn align8(len: usize) -> usize {
    len.next_multiple_of(8)
}

/// One instruction, pointing into the message's shared byte region.
#[derive(Clone, Copy, Debug)]
pub struct Instruction<'a> {
    /// Index into the account keys of the program being invoked.
    pub program_id_index: u8,
    /// Indices into the account keys, in the order the program expects them.
    pub accounts: &'a [u8],
    /// Opaque instruction data.
    pub data: &'a [u8],
}

/// One address lookup table reference, carried unresolved.
///
/// Resolving these needs the lookup tables' account state, which a shred pipeline does not have.
#[derive(Clone, Copy, Debug)]
pub struct AddressLookup<'a> {
    /// The lookup table account.
    pub account_key: &'a [u8; 32],
    /// Indices loaded as writable.
    pub writable_indexes: &'a [u8],
    /// Indices loaded as readonly.
    pub readonly_indexes: &'a [u8],
}

/// A decoded transaction, borrowing the bytes it was read from.
#[derive(Clone, Copy, Debug)]
pub struct Transaction<'a> {
    bytes: &'a [u8],
    signatures_at: usize,
    account_keys_at: usize,
    blockhash_at: usize,
    instruction_table_at: usize,
    instruction_bytes_at: usize,
    lookup_table_at: usize,
    lookup_bytes_at: usize,
    raw_tx_at: usize,
}

impl<'a> Transaction<'a> {
    /// Reads a transaction, validating that every section is present and every offset in range.
    pub fn read(bytes: &'a [u8]) -> Result<Self> {
        if bytes.len() < TX_HEADER_LEN {
            return Err(Error::Truncated {
                what: "transaction header",
                need: TX_HEADER_LEN,
                have: bytes.len(),
            });
        }
        let u16_at = |at: usize| u16::from_le_bytes([bytes[at], bytes[at + 1]]) as usize;
        let u32_at =
            |at: usize| u32::from_le_bytes(bytes[at..at + 4].try_into().expect("in range")) as usize;

        let instruction_bytes_len = u32_at(76);
        let lookup_bytes_len = u32_at(80);
        let account_count = u16_at(92);
        let instruction_count = u16_at(94);
        let signature_count = bytes[96] as usize;
        let lookup_count = bytes[98] as usize;

        let mut at = TX_HEADER_LEN;
        let mut take = |len: usize| -> Result<usize> {
            let start = at;
            at = at.checked_add(len).ok_or(Error::Truncated {
                what: "transaction",
                need: usize::MAX,
                have: bytes.len(),
            })?;
            if at > bytes.len() {
                return Err(Error::Truncated {
                    what: "transaction",
                    need: at,
                    have: bytes.len(),
                });
            }
            Ok(start)
        };

        let signatures_at = take(signature_count * 64)?;
        let account_keys_at = take(account_count * 32)?;
        let blockhash_at = take(32)?;
        let instruction_table_at = take(instruction_count * INSTRUCTION_ENTRY_LEN)?;
        let instruction_bytes_at = take(align8(instruction_bytes_len))?;
        let lookup_table_at = take(lookup_count * LOOKUP_ENTRY_LEN)?;
        let lookup_bytes_at = take(align8(lookup_bytes_len))?;
        let raw_tx_at = take(u32_at(84))?;

        // Table entries are offsets into their own byte region, so they are checked once here and
        // every accessor below can slice without a bounds test.
        for index in 0..instruction_count {
            let entry = instruction_table_at + index * INSTRUCTION_ENTRY_LEN;
            let accounts_end = u16_at(entry + 2) + u16_at(entry + 4);
            let data_end = u32_at(entry + 8) + u32_at(entry + 12);
            if accounts_end > instruction_bytes_len || data_end > instruction_bytes_len {
                return Err(Error::BadOffset {
                    what: "instruction",
                    index,
                });
            }
        }
        for index in 0..lookup_count {
            let entry = lookup_table_at + index * LOOKUP_ENTRY_LEN;
            let writable_end = u16_at(entry + 32) + u16_at(entry + 34);
            let readonly_end = u16_at(entry + 36) + u16_at(entry + 38);
            if writable_end > lookup_bytes_len || readonly_end > lookup_bytes_len {
                return Err(Error::BadOffset {
                    what: "lookup",
                    index,
                });
            }
        }

        Ok(Self {
            bytes,
            signatures_at,
            account_keys_at,
            blockhash_at,
            instruction_table_at,
            instruction_bytes_at,
            lookup_table_at,
            lookup_bytes_at,
            raw_tx_at,
        })
    }

    /// The bytes this transaction was read from.
    #[inline]
    pub const fn as_bytes(&self) -> &'a [u8] {
        self.bytes
    }

    #[inline]
    fn u16_at(&self, at: usize) -> u16 {
        u16::from_le_bytes([self.bytes[at], self.bytes[at + 1]])
    }

    #[inline]
    fn u32_at(&self, at: usize) -> u32 {
        u32::from_le_bytes(self.bytes[at..at + 4].try_into().expect("in range"))
    }

    #[inline]
    fn u64_at(&self, at: usize) -> u64 {
        u64::from_le_bytes(self.bytes[at..at + 8].try_into().expect("in range"))
    }

    /// Slot the transaction landed in.
    #[inline]
    pub fn slot(&self) -> u64 {
        self.u64_at(0)
    }

    /// The slot this one builds on.
    #[inline]
    pub fn parent_slot(&self) -> u64 {
        self.u64_at(8)
    }

    /// When the first shred of the FEC set arrived, on the server's monotonic clock.
    #[inline]
    pub fn rx_ts_ns(&self) -> u64 {
        self.u64_at(16)
    }

    /// When the server finished encoding this message, on the same clock as [`Self::rx_ts_ns`].
    #[inline]
    pub fn emit_ts_ns(&self) -> u64 {
        self.u64_at(24)
    }

    /// How long the server took from first shred to encoded message.
    ///
    /// Both stamps come from one monotonic clock, so this is meaningful; neither can be compared
    /// against a local wall clock.
    #[inline]
    pub fn pipeline_ns(&self) -> u64 {
        self.emit_ts_ns().saturating_sub(self.rx_ts_ns())
    }

    /// Leader assigned to the slot, when the schedule knew one.
    #[inline]
    pub fn leader(&self) -> Option<&'a [u8; 32]> {
        self.has(HAS_LEADER)
            .then(|| self.bytes[32..64].try_into().expect("in range"))
    }

    /// Erasure set this transaction was carried in.
    #[inline]
    pub fn fec_set_index(&self) -> u32 {
        self.u32_at(64)
    }

    /// Index of the entry within the slot.
    #[inline]
    pub fn entry_index(&self) -> u32 {
        self.u32_at(68)
    }

    /// Index of the transaction within its entry.
    #[inline]
    pub fn tx_index(&self) -> u32 {
        self.u32_at(72)
    }

    /// Which ingress source delivered the shreds.
    #[inline]
    pub fn source_id(&self) -> u16 {
        self.u16_at(88)
    }

    /// Raw header flags. Prefer the named accessors.
    #[inline]
    pub fn flags(&self) -> u16 {
        self.u16_at(90)
    }

    #[inline]
    fn has(&self, bits: u16) -> bool {
        self.flags() & bits == bits
    }

    /// Whether this is a simple vote transaction.
    #[inline]
    pub fn is_vote(&self) -> bool {
        self.has(IS_VOTE)
    }

    /// Whether the FEC set had to be reconstructed from parity shreds.
    #[inline]
    pub fn recovered(&self) -> bool {
        self.has(RECOVERED)
    }

    /// How much the server could vouch for this transaction.
    pub fn verification(&self) -> Verification {
        if self.has(VERIFIED) {
            return Verification::Verified;
        }
        match (self.flags() >> REASON_SHIFT) & REASON_MASK {
            1 => Verification::UnknownLeader,
            2 => Verification::StaleSchedule,
            _ => Verification::Disabled,
        }
    }

    /// Number of account keys carried in the message.
    #[inline]
    pub fn account_count(&self) -> usize {
        self.u16_at(92) as usize
    }

    /// Number of instructions.
    #[inline]
    pub fn instruction_count(&self) -> usize {
        self.u16_at(94) as usize
    }

    /// Number of signatures.
    #[inline]
    pub fn signature_count(&self) -> usize {
        self.bytes[96] as usize
    }

    /// Message version, or [`MESSAGE_VERSION_LEGACY`] for a legacy message.
    #[inline]
    pub fn message_version(&self) -> u8 {
        self.bytes[97]
    }

    /// Number of address lookup table references.
    #[inline]
    pub fn lookup_count(&self) -> usize {
        self.bytes[98] as usize
    }

    /// Signatures the message requires.
    #[inline]
    pub fn required_signatures(&self) -> u8 {
        self.bytes[99]
    }

    /// Signed accounts that are readonly.
    #[inline]
    pub fn readonly_signed(&self) -> u8 {
        self.bytes[100]
    }

    /// Unsigned accounts that are readonly.
    #[inline]
    pub fn readonly_unsigned(&self) -> u8 {
        self.bytes[101]
    }

    /// The transaction's first signature — its id.
    #[inline]
    pub fn signature(&self) -> Option<&'a [u8; 64]> {
        (self.signature_count() > 0)
            .then(|| self.bytes[self.signatures_at..self.signatures_at + 64].try_into())
            .and_then(std::result::Result::ok)
    }

    /// Every signature, in order.
    pub fn signatures(&self) -> impl ExactSizeIterator<Item = &'a [u8; 64]> {
        let bytes = self.bytes;
        let at = self.signatures_at;
        (0..self.signature_count())
            .map(move |i| bytes[at + i * 64..at + i * 64 + 64].try_into().expect("in range"))
    }

    /// One account key, by index.
    pub fn account_key(&self, index: usize) -> Option<&'a [u8; 32]> {
        if index >= self.account_count() {
            return None;
        }
        let at = self.account_keys_at + index * 32;
        self.bytes[at..at + 32].try_into().ok()
    }

    /// Every account key, in the order the message declared them.
    pub fn account_keys(&self) -> impl ExactSizeIterator<Item = &'a [u8; 32]> {
        let bytes = self.bytes;
        let at = self.account_keys_at;
        (0..self.account_count())
            .map(move |i| bytes[at + i * 32..at + i * 32 + 32].try_into().expect("in range"))
    }

    /// The blockhash the transaction was signed against.
    #[inline]
    pub fn recent_blockhash(&self) -> &'a [u8; 32] {
        self.bytes[self.blockhash_at..self.blockhash_at + 32]
            .try_into()
            .expect("in range")
    }

    /// One instruction, by index.
    pub fn instruction(&self, index: usize) -> Option<Instruction<'a>> {
        if index >= self.instruction_count() {
            return None;
        }
        let entry = self.instruction_table_at + index * INSTRUCTION_ENTRY_LEN;
        let base = self.instruction_bytes_at;
        let accounts_at = base + self.u16_at(entry + 2) as usize;
        let accounts_len = self.u16_at(entry + 4) as usize;
        let data_at = base + self.u32_at(entry + 8) as usize;
        let data_len = self.u32_at(entry + 12) as usize;
        Some(Instruction {
            program_id_index: self.bytes[entry],
            accounts: &self.bytes[accounts_at..accounts_at + accounts_len],
            data: &self.bytes[data_at..data_at + data_len],
        })
    }

    /// Every instruction, in execution order.
    pub fn instructions(&self) -> impl ExactSizeIterator<Item = Instruction<'a>> {
        let this = *self;
        (0..self.instruction_count()).map(move |i| this.instruction(i).expect("index in range"))
    }

    /// Every address lookup table reference, unresolved.
    pub fn lookups(&self) -> impl ExactSizeIterator<Item = AddressLookup<'a>> {
        let this = *self;
        (0..self.lookup_count()).map(move |i| {
            let entry = this.lookup_table_at + i * LOOKUP_ENTRY_LEN;
            let base = this.lookup_bytes_at;
            let writable_at = base + this.u16_at(entry + 32) as usize;
            let writable_len = this.u16_at(entry + 34) as usize;
            let readonly_at = base + this.u16_at(entry + 36) as usize;
            let readonly_len = this.u16_at(entry + 38) as usize;
            AddressLookup {
                account_key: this.bytes[entry..entry + 32].try_into().expect("in range"),
                writable_indexes: &this.bytes[writable_at..writable_at + writable_len],
                readonly_indexes: &this.bytes[readonly_at..readonly_at + readonly_len],
            }
        })
    }

    /// The original transaction bytes, when the server was asked to include them.
    ///
    /// Off by default: the decoded form carries the same information, and repeating the bytes
    /// roughly doubles the stream's bandwidth.
    pub fn raw_tx(&self) -> Option<&'a [u8]> {
        self.has(HAS_RAW_TX)
            .then(|| &self.bytes[self.raw_tx_at..self.raw_tx_at + self.u32_at(84) as usize])
    }
}
